package/MDAnalysis/analysis/bat.py

```diff
@@ -0,0 +1,503 @@
+            the class. The molecule will be whole opposed to wrapped around a
+            periodic boundary.
+        """
+        # Split the bat vector into more convenient variables
+        origin = bat[:3]
+        (phi, theta, omega) = bat[3:6]
+        (r01, r12, a012) = bat[6:9]
+        n_torsions = (self._ag.n_atoms - 3)
+        bonds = bat[9:n_torsions + 9]
+        angles = bat[n_torsions + 9:2 * n_torsions + 9]
+        torsions = bat[2 * n_torsions + 9:]
+        # When appropriate, convert improper to proper torsions
+        shift = torsions[self._primary_torsion_indices]
+        shift[self._unique_primary_torsion_indices] = 0.
+        torsions += shift
+        # Wrap torsions to between -np.pi and np.pi
+        torsions = ((torsions + np.pi) % (2 * np.pi)) - np.pi
+
+        # Set initial root atom positions based on internal coordinates
+        p0 = np.array([0., 0., 0.])
+        p1 = np.array([0., 0., r01])
+        p2 = np.array([r12 * np.sin(a012), 0., r01 - r12 * np.cos(a012)])
+
+        # Rotate the third atom by the appropriate value
+        co = np.cos(omega)
+        so = np.sin(omega)
+        # $R_Z(\omega)$
+        Romega = np.array([[co, -so, 0], [so, co, 0], [0, 0, 1]])
+        p2 = Romega.dot(p2)
+        # Rotate the second two atoms to point in the right direction
+        cp = np.cos(phi)
+        sp = np.sin(phi)
+        ct = np.cos(theta)
+        st = np.sin(theta)
+        # $R_Z(\phi) R_Y(\theta)$
+        Re = np.array([[cp * ct, -sp, cp * st], [ct * sp, cp, sp * st],
+                       [-st, 0, ct]])
+        p1 = Re.dot(p1)
+        p2 = Re.dot(p2)
+        # Translate the first three atoms by the origin
+        p0 += origin
+        p1 += origin
+        p2 += origin
+
+        XYZ = np.zeros((self._ag.n_atoms, 3))
+        XYZ[self._root_XYZ_inds[0]] = p0
+        XYZ[self._root_XYZ_inds[1]] = p1
+        XYZ[self._root_XYZ_inds[2]] = p2
+
+        # Place the remaining atoms
+        for ((a0,a1,a2,a3), r01, angle, torsion) \
+            in zip(self._torsion_XYZ_inds, bonds, angles, torsions):
+
+            p1 = XYZ[a1]
+            p3 = XYZ[a3]
+            p2 = XYZ[a2]
+
+            sn_ang = np.sin(angle)
+            cs_ang = np.cos(angle)
+            sn_tor = np.sin(torsion)
+            cs_tor = np.cos(torsion)
+
+            v21 = p1 - p2
+            v21 /= np.sqrt(np.sum(v21 * v21))
+            v32 = p2 - p3
+            v32 /= np.sqrt(np.sum(v32 * v32))
+
+            vp = np.cross(v32, v21)
+            cs = np.sum(v21 * v32)
+            if abs(cs) > 1:
+                print('cos ', cs)
+
+            sn = np.sqrt(max(1.0 - cs * cs, 0.0000000001))
+            vp = vp / sn
+            vu = np.cross(vp, v21)
+
+            XYZ[a0] = p1 + \
+              r01*(vu*sn_ang*cs_tor + vp*sn_ang*sn_tor - v21*cs_ang)
+        return XYZ
+
+    def getAtomGroup(self):
+        """Returns the atomgroup
+        """
+        return self._ag
```

Comment:
We generally don't have setters/getters. Instead we use properties. If you want the ag as a "read-only" attribute:
```suggestion
    @property
    def atoms(self):
        """The atomgroup for which BAT are computed (read-only property)"""
        return self._ag
```